### DISCUSSION TITLE
Color "noise": Blending similar colors

### DISCUSSION THREAD

**Discussion by mtrutsch** [ImageMagick/ImageMagick#7785] [General] [answered]
Hello,

I would like to combine similar colors to one color. For example, in the following image, I would like to handle different hues of yellow, so that it would be just one color (no matter what color). I know that I could threshold the colors, however, I do not know what color will be in the background.

Suggestions would be greatly appreciated. Thank you.
![sampleImg](https://github.com/user-attachments/assets/805af6d0-6cb4-4687-9d3e-f3555841d302)

MTX :-)

**Comment by fmw42** [comment]
Convert to HSV colorspace. Then separate HSV channels. Then find the range of hues for the yellow region and the middle of that range (use some color picker for that). Then use -fuzz XX% -fill middle_color -opaque middle_color.  Where XX% is the half range of color values expressed as percent.  I get approximately a range of 8-bit colors from 37 to 43. So 6/56=2.4%  (you may need to increase this a little) with a middle value of 40 as 8-bit or 40/256=15.6%.  Adjust as needed to fill all the yellow value in the hue image as needed as applied to the hue channel.  Then combine the new hue with the old satuation and value channels.  Then convert that back to sRGB colorspace.

**Reply by mtrutsch** [reply]
Thank you, fmw42, for your reply. Unfortunately, I need this to be done automatically, without human intervention.

**Comment by snibgo** [comment, marked as answer]
"-kmeans" may be useful here. For example:
```
magick mtr.png -kmeans 4 x.png
```
The result x.png is:
![x](https://github.com/user-attachments/assets/e57c92b6-5a18-4047-8ec4-c9cee6fa5bd6)

**Reply by mtrutsch** [reply]
Thank you. This great.

**Reply by fmw42** [reply]
What is your IM version and OS/Platform?

**Reply by mtrutsch** [reply]
magick imgIN.png -kmeans **2** imgOUT.png

2 in -kmeans 2 works best in my case.

**Reply by mtrutsch** [reply]
I am using a RPi that does not have magick available (tried updating imagemagick with sudo apt install imagemagick and did not help). However, there is magick on my macOS computer and it worked well.
![imgOUT](https://github.com/user-attachments/assets/de82146d-1e12-43c6-872b-725051d5e701)

**Reply by fmw42** [reply]
But I thought you wanted to keep the other non-yellow colors? Is that not the case?

**Reply by mtrutsch** [reply]
Sorry. I should have mentioned that I do not need other colors. I could live with them, however, it is better if I only have two colors as shown above. :-) Thank you for your help.

**Comment by fmw42** [comment]
How would you know where to separate the yellow from the orange, if one tries to automatically determine the range of yellow?  Can you predefine a tolerance?

**Reply by mtrutsch** [reply]
At this point, I cannot define a tolerance. However, a two color solution works great. :-)

**Comment by mtrutsch** [comment]
Right now, I am trying to figure out how come magick is not available on RPi (bookworm with the latest updates/upgrades).

**Comment by fmw42** [comment]
`magick img.png -kmeans 3 y.png`. Looks reasonable to me as you mostly have 3 colors.

![y](https://github.com/user-attachments/assets/ba5757a6-c25c-4a67-8fc4-2ea2b5843b14)


Alternately, you could make a mask from the kmeans result and use it to composite between the original and the kmeans images

```
magick img.png -alpha off -write mpr:img -kmeans 2 \
\( +clone -auto-threshold otsu \) \
mpr:img -compose over -composite result.png

```

![result](https://github.com/user-attachments/assets/b676bd6d-9ea1-499b-b387-1149aa25994e)


Unfortunately it just gives some yellow haze and does not get the other colors. But you might use -morphology to dilate the white in the mask to reduce the black.

**Reply by mtrutsch** [reply]
This looks useful too. Thank you. :-)

**Comment by mtrutsch** [comment]
By the way, for those who do need to install the latest Imagemagick on their Raspberry Pi, check out the following URL: [https://www.tecmint.com/install-imagemagick-on-debian-ubuntu]. sudo apt install imagemagick does not install the latest version of Imagemagick (7.x+) that comes with 'magick'.
